include/dlaf/eigensolver/tridiag_solver/mc.h

```diff
@@ -0,0 +1,241 @@
+/// │        │      │    │       │         ──────┤    │       │               │    │
+/// ├──────┬─┴──────┤ Qd │       ├───────        │    │       ├───────────────┤ Qd │
+/// │      │        │    │   X   │          U2'  │    │   =   │               │    │
+/// │      │        │    │       │               │    │       │    Q2'xU2'    │    │
+/// │      │  Q2'   │    │       ├───────────────┼────┤       │               │    │
+/// │      │        │    │       │               │ I  │       │               │    │
+/// │      │        │    │       │               │    │       │               │    │
+/// └──────┴────────┴────┘       └───────────────┴────┘       └───────────────┴────┘
+///
+/// Note:
+/// 1. U1' and U2' may overlap (in practice they almost always do)
+/// 2. The overlap between U1' and U2' matches the number of shared columns between Q1' and Q2'
+/// 3. The overlap region is due to deflation via Givens rotations of a column vector from Q1 with a
+///    column vector of Q2.
+template <class T>
+void TridiagSolver<Backend::MC, Device::CPU, T>::call(Matrix<T, Device::CPU>& mat_trd,
+                                                      Matrix<T, Device::CPU>& d,
+                                                      Matrix<T, Device::CPU>& mat_ev) {
+  // Cuppen's decomposition
+  std::vector<pika::shared_future<T>> offdiag_vals = cuppensDecomposition(mat_trd);
+
+  // Solve with stedc for each tile of `mat_trd` (nb x 2) and save eigenvectors in diagonal tiles of
+  // `mat_ev` (nb x nb)
+  solveLeaf(mat_trd, mat_ev);
+
+  // Offload the diagonal from `mat_trd` to `d`
+  offloadDiagonal(mat_trd, d);
+
+  // Auxiliary matrix used for the D&C algorithm
+  const matrix::Distribution& distr = mat_ev.distribution();
+  WorkSpace<T> ws = initWorkSpace<T>(distr);
+
+  // Each triad represents two subproblems to be merged
+  for (auto [i_begin, i_split, i_end] : generateSubproblemIndices(distr.nrTiles().rows())) {
+    mergeSubproblems(i_begin, i_split, i_end, ws, offdiag_vals[to_sizet(i_split)], d, mat_ev);
+  }
+}
+
+/// ---- ETI
+#define DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(KWORD, BACKEND, DEVICE, DATATYPE) \
+  KWORD template struct TridiagSolver<BACKEND, DEVICE, DATATYPE>;
+
+DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, float)
+DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, double)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, std::complex<float>)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::MC, Device::CPU, std::complex<double>)
+
+#ifdef DLAF_WITH_CUDA
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, float)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, double)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, std::complex<float>)
+// DLAF_TRIDIAGONAL_EIGENSOLVER_ETI(extern, Backend::GPU, Device::GPU, std::complex<double>)
```

Comment:
To be enabled in a separate PR or this PR?

Reply:
This will be enabled in a separate PR.